File: lollys-lotto-rust-sdk/src/instructions/create_user_metadata.rs
use crate::instructions::*;

pub fn create_user_metadata(
    user: &Pubkey,
    user_metadata: &Pubkey,
    usdc_mint: &Pubkey,
    user_rewards_vault: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = lollys_lotto::instruction::CreateUserMetadata.data();

    let accounts = lollys_lotto::accounts::CreateUserMetadata{
        user: *user,
        user_metadata: *user_metadata,
        usdc_mint: *usdc_mint,
        user_rewards_vault: *user_rewards_vault,
        event_emitter: *event_emitter,
        token_program: token::ID,
        associated_token_program: associated_token::ID,
        system_program: system_program::ID,
    }.to_account_metas(None);
    Instruction {
        program_id: lollys_lotto::ID,
        accounts,
        data,
    }
}

File: lollys-lotto-ts-sdk/src/codegen/instructions/createUserMetadata.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CreateUserMetadataAccounts {
  user: PublicKey
  userMetadata: PublicKey
  /** Mint address of the USDC token */
  usdcMint: PublicKey
  userRewardsVault: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
}

export function createUserMetadata(
  accounts: CreateUserMetadataAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.user, isSigner: true, isWritable: true },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    { pubkey: accounts.usdcMint, isSigner: false, isWritable: false },
    { pubkey: accounts.userRewardsVault, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([56, 247, 226, 15, 184, 166, 38, 102])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
